lib/que/worker.rb

```diff
@@ -1,184 +1,123 @@
 # frozen_string_literal: true
 
-require 'monitor'
+require "benchmark"
 
 module Que
   class Worker
-    # Each worker has a thread that does the actual work of running jobs.
-    # Since both the worker's thread and whatever thread is managing the
-    # worker are capable of affecting the worker's state, we need to
-    # synchronize access to it.
-    include MonitorMixin
+    # Defines the time a worker will wait before checking Postgres for its next job
+    WAKE_INTERVAL = 5
+    DEFAULT_QUEUE = ''
+    JOB_INSTANCE_FIELDS = %i[queue priority run_at job_id].freeze
 
-    attr_reader :thread, :state, :queue
-
-    def initialize(queue = '')
-      super() # For MonitorMixin.
+    def initialize(queue: DEFAULT_QUEUE, wake_interval: WAKE_INTERVAL)
       @queue  = queue
-      @state  = :working
-      @thread = Thread.new { work_loop }
-      @thread.abort_on_exception = true
+      @wake_interval = wake_interval
+      @stop = false
     end
 
-    def alive?
-      !!@thread.status
-    end
+    def work_loop
+      return if @stop
 
-    def sleeping?
-      synchronize { _sleeping? }
-    end
+      loop do
+        case work
+        when :job_not_found, :error then sleep(@wake_interval)
+        when :job_worked then nil # immediately find a new job to work
+        end
 
-    def working?
-      synchronize { @state == :working }
+        break if @stop
+      end
     end
 
-    def wake!
-      synchronize do
-        if sleeping?
-          # Have to set the state here so that another thread checking
-          # immediately after this won't see the worker as asleep.
-          @state = :working
-          @thread.wakeup
-          true
+    def work
+      Que.adapter.checkout do
+        with_locked_job do |job|
+          return :job_not_found if job.nil?
+
+          # Check that the job hasn't just been worked by another worker (it's possible to
+          # lock a job that's just been destroyed because pg locks don't obey MVCC). If it
+          # has been worked, act as if we've worked it.
+          return :job_worked unless job_exists?(job)
+
+          log_keys = job.values_at("id", "priority", "args", "queue")
+
+          begin
+            Que.logger&.info(
+              log_keys.merge(
+                event: "job_begin",
+                msg: "Job acquired, beginning work",
+              )
+            )
+
+            klass = class_for(job[:job_class])
+            # TODO: _run -> run_and_destroy(*job[:args])
+            duration = Benchmark.measure { klass.new(job)._run }.real
```

Comment:
FYI @jacobpgn, underneath the hook this [uses `Process::CLOCK_MONOTONIC`](https://github.com/ruby/ruby/blob/trunk/lib/benchmark.rb#L292). Thought you'd find that interesting!

Reply:
😻